blog/_src/posts/2017-11-XX-adding-refinement-types.md

```diff
@@ -0,0 +1,292 @@
+
+```racket
+(Refine [i : Integer] (<= 0 i (- (vector-length v) 1)))
+```
+
+This refinement type describes values `i` of type `Integer` such that `(<= 0 i (- (vector-length v) 1))` produces `#true`. The documentation for [Refine](http://docs.racket-lang.org/search/index.html?q=Refine) describes precisely what propositions can be used in refinements. Today it includes logical statements about the types of terms, standard logical combinators (e.g. and, or), and linear integer constraints.
+
+## Dependent function types
+
+By combining refinement types with dependent function types (i.e. functions whose domain and range may depend on argument and/or program values), we can now give a more precise function type for many basic vector operations, e.g.:
+
+```racket
+(: safe-vector-ref 
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : (v) (Refine [i : Integer]
+                                  (<= 0 i (- (vector-length v) 1)))])
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+Here we have stated that `safe-vector-ref`'s second argument (`n`) depends on its first argument (`v`) and that the type of `n` is an `Integer` which is a valid index for `v`.
+
+If we try and use `safe-vector-ref` with a vector and index which Typed Racket cannot prove satisfy the respective types:
+
+```racket
+(safe-vector-ref (vector 0 1) 2)
+```
+
+Typed Racket will report an error while type checking:
+
+```
+Type Checker: type mismatch
+  expected: (Refine
+             (i : Integer)
+             (and (<= i (- (vector-length v) 1)) (<= 0 v)))
+  given: (Refine (z : Positive-Byte) (= 2 z)) in: 2
+```
+
+When a refinement on an argument's type may also be written conveniently as a "precondition", e.g.:
+
+```racket
+(: safe-vector-ref
+   (All (A) (-> ([v : (Vectorof A)]
+                 [n : Natural])
+                #:pre (v n) (< n (vector-length v))
+                A)))
+(define safe-vector-ref vector-ref)
+```
+
+Typed Racket may be able to provide more detailed feedback when the arguments are of the correct type but the precondition cannot be proved:
+
+```
+Type Checker: could not apply function;
+ unable to prove precondition
+  precondition: (<= 2 (- (vector-length v) 1)) 
+  in: (safe-vector-ref (vector 0 1) 2)
+```
+
+We can read the precondition specification above as follows: the precondition for the function depends are arguments `v` and `n`, and it requires that Typed Racket can prove `(<= 0 n (- (vector-length v) 1))` would produce `#true` at all call sites.
+
+## Programming with refinement types
+
+_Note: these examples should work on any Racket builds from 3 November or later.
+To use with Racket v6.11, an additional small snippet of code (included at the end of
+this post) is required._
+
+When programming with refinement types, use the `#:with-refinements` option on the `#lang` line to tell Typed Racket to more liberally track detailed logical information about program terms:
```

Comment:
Using "liberally" here seems correct in the sense of "in generous amounts" like "sprinkle liberally with salt" but it's at odds with "liberally" meaning "in an imprecise or loose manner," which the opposite of what you're trying to convey here.  I'd pick a different word that just has the first sense.